sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -540,10 +540,11 @@ object QueryExecution {
   }
 
   /**
-   * Converts asserts, null pointer exceptions to internal errors.
+   * Converts asserts, null pointer exceptions and scala match errors to internal errors.
    */
   private[sql] def toInternalError(msg: String, e: Throwable): Throwable = e match {
-    case e @ (_: java.lang.NullPointerException | _: java.lang.AssertionError) =>
+    case e @ (_: java.lang.NullPointerException | _: java.lang.AssertionError |
+              _: scala.MatchError) =>
```

Comment:
we can replace this case match by moving the logic to a new method to make the code more reable
```
private def isInternalError(e: Throwable): Boolean = e match {
  case _: java.lang.NullPointerException => true
  case _: _: java.lang.AssertionError => true
  ...
  case _ => false
}
```

then the code here
```
if (isInternalError(e)) {
  SparkException.internalError...
} else {
  e
}
```